src/components/graph-sonification.tsx

```diff
@@ -0,0 +1,174 @@
+import { codapInterface, getAllItems, IResult } from "@concord-consortium/codap-plugin-api";
+import React, { useEffect, useRef } from "react";
+import * as Tone from "tone";
+import { CodapItem } from "../types";
+
+interface IProps {
+  graphToSonify: string;
+  onResetGraphToSonify: () => void;
+}
+
+export const GraphSonification = ({ graphToSonify, onResetGraphToSonify }: IProps) => {
+  const synthRef = useRef<Tone.PolySynth | null>(null);
+  const pannerRef = useRef<Tone.Panner | null>(null);
+  const frameIdRef = useRef<number | null>(null);
+  const currentGraphIdRef = useRef<string | null>(null);
+
+  if (!pannerRef.current) {
+    pannerRef.current = new Tone.Panner(0).toDestination();
+  }
+
+  if (!synthRef.current) {
+    synthRef.current = new Tone.PolySynth().connect(pannerRef.current);
+  }
+
+  const removeROIAdornment = async (graphId: string) => {
+    await codapInterface.sendRequest({
+      action: "delete",
+      resource: `component[${graphId}].adornment`,
+      values: {
+        type: "Region of Interest"
+      }
+    });
+  };
+
+  useEffect(() => {
+    const fetchGraphData = async () => {
+      const res = await codapInterface.sendRequest({action: "get", resource: `component[${graphToSonify}]`}) as IResult;
+      const graphDetails = res.values;
+      if (!graphDetails) return;
+
+      if (currentGraphIdRef.current && currentGraphIdRef.current !== graphDetails.id) {
+        await removeROIAdornment(currentGraphIdRef.current);
+      }
+
+      currentGraphIdRef.current = graphDetails.id;
+
+      await codapInterface.sendRequest({
+        action: "create",
+        resource: `component[${graphDetails.id}].adornment`,
+        values: {
+          type: "Region of Interest",
+          primary: { "position": 0, "extent": 0.05 },
+          secondary: { "position": 0, "extent": "100%" }
+        }
+      });
+
+      const pitchAttr = graphDetails.yAttributeName;
+      const timeAttr = graphDetails.xAttributeName;
+
+      const allItemsRes = await getAllItems(graphDetails.dataContext);
+      const allItems = allItemsRes.values;
```

Comment:
Would this be a problem if there are many thousands of points?

Reply:
It doesn't seem to be a problem in terms of getting the data. We tried it out with a dataset containing several thousand cases and there didn't seem to be any performance issues. The actual sound may leave something to be desired, but I think that's something the project folks will want to experience and think about before we work out how to possibly make it work better.